src/main/kotlin/org/rust/ide/intentions/ExtractEnumVariantIntention.kt

```diff
@@ -179,92 +197,129 @@ class ExtractEnumVariantIntention : RsElementBaseIntentionAction<ExtractEnumVari
-    override val typeReferences: List<RsTypeReference> = fields.tupleFieldDeclList.map { it.typeReference }
+    protected fun formatFields(): String {
+        val fields = toBeReplaced.copy()
 
-    override fun createStruct(name: String, typeParameters: String, whereClause: String): RsStructItem =
-        factory.createStruct("struct $name$typeParameters${fields.text}$whereClause;")
+        val decls: List<RsFieldDecl> = when (fields) {
+            is RsBlockFields -> fields.namedFieldDeclList
+            is RsTupleFields -> fields.tupleFieldDeclList
+            else -> error("unreachable")
+        }
 
-    override fun replaceUsage(element: PsiElement, name: String) {
-        if (replaceTuplePattern(element, name)) return
-        replaceTupleCall(element, name)
-    }
+        val declsWithoutVis = decls.filter { it.vis == null }
+        if (declsWithoutVis.isNotEmpty()) {
+            val pub = factory.createPub()
+            val whitespace = factory.createWhitespace(" ")
+            for (field in declsWithoutVis) {
+                when (field) {
+                    is RsNamedFieldDecl -> {
+                        field.addBefore(pub, field.identifier)
+                    }
+                    is RsTupleFieldDecl -> {
+                        val typeReference = field.typeReference
+                        field.addBefore(pub, typeReference)
+                        if (typeReference.refLikeType != null) {
+                            field.addBefore(whitespace, typeReference)
+                        }
+                    }
+                }
+            }
+        }
 
-    private fun replaceTupleCall(element: PsiElement, name: String): Boolean {
-        val parent = element.ancestorStrict<RsCallExpr>() ?: return false
-        val call = factory.createFunctionCall(name, parent.valueArgumentList.text.trim('(', ')'))
-        val binding = factory.createFunctionCall(element.text, listOf(call))
-        parent.replace(binding)
-        return true
+        return fields.text
     }
+}
 
-    private fun replaceTuplePattern(element: PsiElement, name: String): Boolean {
-        val parent = element.ancestorStrict<RsPatTupleStruct>() ?: return false
-        val binding = factory.createPatTupleStruct(name, parent.patList)
+private class TupleVariant(fields: RsTupleFields) : VariantElement(fields) {
+    override val typeReferences: List<RsTypeReference> = fields.tupleFieldDeclList.map { it.typeReference }
 
-        for (pat in parent.patList) {
-            val comma = pat.nextSibling
-            pat.delete()
-            if (comma.text == ",") {
-                comma.delete()
+    override fun createStruct(vis: String?, name: String, typeParameters: String, whereClause: String): RsStructItem {
+        val formattedVis = if (vis == null) "" else "$vis "
+        val fields = formatFields()
+        return factory.createStruct("${formattedVis}struct $name$typeParameters$fields$whereClause;")
```

Comment:
`org.rust.ide.intentions.TupleVariant#createStruct` and `org.rust.ide.intentions.StructVariant#createStruct` looks very similar
Maybe it's better not to override it in each inheritor, move all code in base class and make `when` for struct text?